### PR TITLE
Fast rise drop alerts

### PR DESCRIPTION
**Pull request by opsb**
Adds fast drop and fast rise alerts, by default they're triggered when delta exceeds 20mg/dl. 

One thing I'm not sure about is how to handle the default value depending on the user's setting for blood glucose value units (mgdl/mmol). There's likely a quick fix for this scenario but it got me wondering if there should be a value type to represent the BGValue and to this end I've made a proposal at https://github.com/JohanDegraeve/xdripswift/issues/67.

### CODE REVIEW

**Comment by JohanDegraeve** [comment]
regarding the change, the only things that the new alerts will appear below the battery low alert,
As mentioned in a comment of AlertKind
https://github.com/JohanDegraeve/xdripswift/blob/a2a34479df821d0b3888389ba41ad69150cf7aab/xdrip/Managers/Alerts/AlertKind.swift#L5

changing the order would be better from ui point of view, but would require some migration

what do you think ?

**Comment by opsb** [comment]
I've updated the AlertsManager so that only a single alert is raised from `[.fastdrop, .verylow, .low, .fastrise, .veryhigh, .high, .calibration, .batterylow]`

**Comment by opsb** [comment]
> regarding the change, the only things that the new alerts will appear below the battery low alert,
> As mentioned in a comment of AlertKind
> https://github.com/JohanDegraeve/xdripswift/blob/a2a34479df821d0b3888389ba41ad69150cf7aab/xdrip/Managers/Alerts/AlertKind.swift#L5
> 
> changing the order would be better from ui point of view, but would require some migration
> 
> what do you think ?

Yeah, I agree it would be nicer if the alerts were grouped with the other BG alerts. I tried changing to

    ```
    case fastdrop = 0
    case verylow = 1
    case low = 2
    case high = 3
    case veryhigh = 4
    case fastrise = 5
    case missedreading = 6
    case calibration = 7
    case batterylow = 8
```

but I got some weird errors with alert messages being different alerts (presumably because they're looked up by index somewhere) so I backed away.

One option would be to explicitly list the preferred order of the alerts for the UX, I couldn't think of a way to do this with exhaustive checking but perhaps that isn't so important.

**Comment by opsb** [comment]
I've changed the fast drop / fast rise delta triggers to 10mg/dl (matches the default in spike which I was pretty happy with)

**Comment by JohanDegraeve** [comment]
I think we could do a kind of migration.
the alertkind is stored in the coredata as Int16
https://github.com/JohanDegraeve/xdripswift/blob/master/xdrip/Core%20Data/classes/AlertEntry%2BCoreDataClass.swift#L24

There should be at least one alertentry for every kind.
If there's none for values 7 and 8, it means it's an 'old' database, all entries would values 6 would need to be changed to 8, 5 to 7, 4 to 6

Do you think that's feasible ?
This migration could  be done here i think https://github.com/JohanDegraeve/xdripswift/blob/master/xdrip/Managers/CoreData/CoreDataManager.swift#L114
Right before completion is called. Calling completion happes later and invokes setting up the transmitter, chart . etc.

I'll have a look at the snooze changes you did

**Comment by opsb** [comment]
> There should be at least one alertentry for every kind.
Given this assumption holds then I don't see any issue with doing a migration. I'm not entirely convinced that we should be using the enum case ordering to define the UX ordering but I'm happy to put it together for this PR.

**Comment by JohanDegraeve** [comment]
it's indeed not the nicest solution with the enum case ordering.

we could add a mapping functions that maps a case to a row number and vice versa

**Comment by JohanDegraeve** [comment]
If you prefer, I'll merge the change later today and add a mapping function.

Op ma 27 jan. 2020 om 08:00 schreef Johan Degraeve <johan.degraeve@gmail.com

**Comment by opsb** [comment]
Happy for you to go ahead, not sure when I'll have time now this week.

I was giving some thought to storing the enum values in coredata as strings rather than index numbers. That way the storage would be immune to order changes. I'm not familiar with coredata though so no idea how much work would be involved so it's just a thought at this point.

File: xdrip/Managers/Alerts/AlertManager.swift [hunk 1]
```diff
@@ -137,27 +137,16 @@ public class AlertManager:NSObject {
...
-                                    // finally let's check the battery level alert
-                                    _ = checkAlertAndFire(alertKind: .batterylow, lastBgReading: lastBgReading, lastButOneBgREading: lastButOneBgREading, lastCalibration: lastCalibration, transmitterBatteryInfo: transmitterBatteryInfo)
-                                }
-                            }
-                        }
-                    }
-                }
-                // set missed reading alert, this will be a future planned alert
+
+                // TODO original implementation ignores isSnoozed for .calibration and .batterylow
```
Comment (JohanDegraeve) [inline xdrip/Managers/Alerts/AlertManager.swift]: to snooze status is checked in function checkAlertAndFire, also for battery and calibration
The additional check for isSnoozed is not there to check if the alert itself is snoozed but if a higher level alert is snoozed.

Example, very low is snoozed, will proceed to next alert with is low. But if user has snoozed verylow, then probably he or she doesn't want to have an alert for low

**Comment by JohanDegraeve** [comment]
not bad for not being used to swift :)

I have a mapping function ready, I'll test further tonight and then check your latest commit

File: xdrip/Managers/Alerts/AlertManager.swift (hunk 1 above)
Comment (opsb) [reply xdrip/Managers/Alerts/AlertManager.swift]: Ah that makes, I've updated the code accordingly

**Comment by JohanDegraeve** [comment]
I've created my own branch named opsb-fast-rise-drop-alerts, just pushed to my repository.
I still need to test the alerts now

**Comment by JohanDegraeve** [comment]
I'm releasing the feature.  I propose to close this pull request.
There's already a list of bugs and enhancement requests, if you think you can give more contributions let me know please

**Comment by opsb** [comment]
Hey @JohanDegraeve, sounds good. I've definitely found the alerts very useful in the last few days. I'll take a look at the issues.

**Pull request closed without merging**